[SPARK-6980] [CORE] [WIP] Akka timeout exceptions indicate which conf controls them

BryanCutler opened the pull request:
First shot at adding a description for an akka timeout that indicates which configuration property it came from

BryanCutler committed:
[SPARK-6980] Akka timeout exceptions indicate which conf controls them

AmplabJenkins commented:
Can one of the admins verify this patch?

BryanCutler commented:
I still pretty new with Scala, so any comments/suggestions is much appreciated!

squito commented:
jenkins, this is ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31255 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31255/consoleFull) for   PR 5741 at commit [`9368e48`](https://github.com/apache/spark/commit/9368e480c469eb0f0d36fbc32af6c9896034d01b).

SparkQA commented:
[Test build #31255 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31255/consoleFull) for   PR 5741 at commit [`9368e48`](https://github.com/apache/spark/commit/9368e480c469eb0f0d36fbc32af6c9896034d01b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ConfiguredTimeout(timeout_duration: FiniteDuration, timeout_description: String = null) `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31255/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/util/AkkaUtils.scala:
```diff
@@ -30,6 +32,48 @@ import org.apache.log4j.{Level, Logger}
+ * @param timeout_duration timeout duration in milliseconds
+ * @param timeout_description description to be displayed in a timeout exception
+ */
+class ConfiguredTimeout(timeout_duration: FiniteDuration, timeout_description: String = null) {
```
scala convention is camelCase, so `timeoutDuration` and `timeoutDescription`.  But in this case, I think `timeout` is a little redundant anyway, so maybe just `duration` and `description`.  I don't think you should have a `null` default for `description`

squito commented on core/src/main/scala/org/apache/spark/util/AkkaUtils.scala:
```diff
@@ -30,6 +32,48 @@ import org.apache.log4j.{Level, Logger}
+   * @return ConfiguredTimeout object
+   */
+  implicit def toConfiguredTimeout(timeout_duration: FiniteDuration): ConfiguredTimeout =
+    new ConfiguredTimeout(timeout_duration)
```
I don't think we want an implicit conversion here.  The point is to *force* you to give a `ConfiguredTimeout` when you call `AkkaUtils.askWithReply`.  That's to make sure that we get the better error messages consistently.

of course, this means you've got more work to do, tracking down each use of `askWithReply`, and putting the right `ConfiguredTimeout`.  Its possible there are some cases which don't have a timeout that is tied to a configuration parameter, so maybe this won't work.  (Or maybe we should change those cases so they are configurable.)

Also, as cool as implicits are, they end up making the code significantly harder to read, so we try to avoid them.  (eg., see http://twitter.github.io/effectivescala/#Types and Generics-Implicits)

squito commented on core/src/main/scala/org/apache/spark/util/AkkaUtils.scala:
```diff
@@ -171,14 +215,20 @@ private[spark] object AkkaUtils extends Logging {
+        case te: TimeoutException =>
+          var msg = te.toString()
+          if (confTimeout.description != null) {
+            msg += " with [" + confTimeout.description + "]"
```
I'd like a little clearer explanation here, like "This timeout is controlled by spark.xxx.yyy.zzz"

squito commented:
Thanks for working on this @BryanCutler!  my major concern at this point is the use of the implicit conversion.  Also we need to add a test to `AkkaUtilsSuite` for `askWithReply`, that the timeout still works, and we get the right error message when there is no response in time.

BryanCutler commented:
Thanks for the quick feedback @squito!  I probably should have explained a little, but I only added the implicit conversion to maintain api compatibility by also allowing a {{FiniteDuration}} input parameter.  If that's not needed then I can definitely remove it.

I'll fix up the code with your suggestions and see what I can do about adding a test case for this.  Hopefully, be able to get it done in the next couple days.

BryanCutler committed:
[SPARK-6980] Updated with comments from PR, added test in AkkaUtilsSuite and new test suite ActorSystemSuite

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #31686 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31686/consoleFull) for   PR 5741 at commit [`d4bb0e9`](https://github.com/apache/spark/commit/d4bb0e9309df6d681eb7a809590cf859e7930701).

BryanCutler commented:
Made another commit with received feedback.  Changes are:
* removed implicit conversion, ConfiguredTimeout must be constructed with a description
* more descriptive message printed when timeout
* added method to create ConfiguredTimeout with properties used in RpcUtils.askTimeout
* added test in AkkaUtilsSuite and new suite ActorSystemSuite to test with an ActorSystem

SparkQA commented:
[Test build #31686 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31686/consoleFull) for   PR 5741 at commit [`d4bb0e9`](https://github.com/apache/spark/commit/d4bb0e9309df6d681eb7a809590cf859e7930701).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `class ConfiguredTimeout(duration: FiniteDuration, description: String) `

 * This patch **adds the following new dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `mesos-0.21.0-shaded-protobuf.jar`

 * This patch **removes the following dependencies:**
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `mesos-0.21.1-shaded-protobuf.jar`
   * `pmml-agent-1.1.15.jar`
   * `pmml-model-1.1.15.jar`
   * `pmml-schema-1.1.15.jar`
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31686/
Test PASSed.

squito commented:
Hi @BryanCutler, thanks for the updates.

I realized that I between the time I dreamed of this in my head, and when I actually created the jira, there was a layer of abstraction put in the RPC layer.   So in addition to `AkkaUtils.askWithReply`, we also want to fix `AkkaRpcEnv.sendWithReply`, as that is where most of the calls are really being made.  As you pointed out, these timeouts are all created by `RpcUtils.askTimeout` -- so how about we just change that method to always return a `ConfiguredTimeout`?

Sorry I spec'ed this poorly in the beginning and to require more work on this, but I feel pretty strongly that I want to make sure we get *all* of the timeouts covered, and the best way to do that is make the compiler do the checks for us.  Also it makes me to want to fix `RpcUtils.lookupTimeout` as well.  Just changing `AkkaUtils.askWithReply` is still useful, so you dont' have to take on the whole thing, but there's no huge time pressure so feel free to keep going on it.

squito commented on streaming/src/test/scala/org/apache/spark/streaming/ActorSystemSuite.scala:
```diff
@@ -0,0 +1,73 @@
+
+    conf.set(shortProp, "1s")
+
+    ssc = new StreamingContext(master, appName, batchDuration)
```
you don't need to make a streamingContext for this test at all, it just involves akka actors.  In fact, I think this test should get moved to AkkaUtilsSuite as well.

BryanCutler replied:
No problem, in my initial tests I created the Actors from the ActorSystem, so I thought that was required.  I'll rework this into AkkaUtilsSuite and shorten the timeouts as you mentioned below.

squito commented on streaming/src/test/scala/org/apache/spark/streaming/ActorSystemSuite.scala:
```diff
@@ -0,0 +1,73 @@
+    val askingActor = actorSystem.actorOf(Props[EchoActor], "AskingActor")
+
+    AkkaUtils.askWithReply[String]("this should echo", askingActor, 1, 0,
+      ConfiguredTimeout.createAskTimeout(conf))
```
you could get around the fact that the conf timeouts have a minimum time of seconds by directly creating your timeout here: `new ConfiguredTimeout(20 milliseconds, "This timeout is configured by foobar")`.  1 second for a test is not the end of the world, but would be nice to get it shorter.  I'd also be sure to have one test where the reply *is* received within the timeout, so that we know we don't always incorrectly throw the exception.

squito commented on streaming/src/test/scala/org/apache/spark/streaming/ActorSystemSuite.scala:
```diff
@@ -0,0 +1,73 @@
+      throw new TestException("required exception not raised during AkkaUtils.askWithReply")
+    } catch {
+      case se: SparkException =>
+        assert(se.getCause().getMessage().contains(shortProp))
```
this can be simplified to:
```
val exc = intercept[SparkException]{
  AkkaUtils.askWithReply[String]("this should timeout", askingActor, 1, 0,
        ConfiguredTimeout(conf, shortProp))
}
assert(exc.getCause().getMessage().contains("..."))
```

BryanCutler replied:
that's a nice trick, thanks!

squito commented on core/src/main/scala/org/apache/spark/util/AkkaUtils.scala:
```diff
@@ -30,6 +34,61 @@ import org.apache.log4j.{Level, Logger}
+ * @param duration timeout duration in milliseconds
+ * @param description description to be displayed in a timeout exception
+ */
+class ConfiguredTimeout(duration: FiniteDuration, description: String) {
```
this need to be scoped at least to `private[spark]`, possibly even tighter.  Same goes for other classes & methods.  We can wait till the rest of the design shakes out a little more before worrying about that too much, though

BryanCutler commented:
Hey @squito, I agree that it would be best to improve on all of these timeouts.  I'll extend the scope to include `AkkaRpcEnv.sendWithReply` and modify `askTimeout` and `lookupTimeout`.  Should we try something to handle the timeouts from using these operators:
`sender ! message`
`sender ? message`
or since this usage is out of the Spark realm, we don't need to worry about it?

zsxwing commented:
Could you take a look at `org.apache.spark.rpc.RpcEndpointRef` and `org.apache.spark.rpc.RpcEnv` and apply `ConfiguredTimeout` there? `AkkaUtils.askWithReply` won't be used after we finish refactoring of the RPC layer.

BryanCutler commented:
Sure, I can do that.  I'll follow up with another PR soon, thanks!

squito commented:
Thanks for the feedback @zsxwing !  and thanks for sticking with this Bryan

squito commented:
@BryanCutler to answer your earlier question about `sender ! message` and `sender ? message`:

`!` doesn't have a timeout, so nothing to change there.  And for other occurrences of `?` -- my guess is that the other refactoring going on will eliminate them.  Wouldn't hurt for you to change them as well so they don't overlooked, but might be more work  and / or redundant w/ the refactoring.  Sorry we're going after a bit of a moving target here :)

hardmettle commented:
@BryanCutler @squito  Also one more confusion I had was how would you differentiate between the constructors (in the apply method) since the description and timeoutProp are both strings.

BryanCutler commented:
Closing, will continue in PR https://github.com/apache/spark/pull/6205
